language/diem-vm/parallel-executor/src/executor.rs

```diff
@@ -0,0 +1,185 @@
+        let infer_result: Vec<_> = {
+            match signature_verified_block
+                .par_iter()
+                .with_min_len(chunks_size)
+                .map(|txn| {
+                    Ok((
+                        self.inferencer.infer_reads(txn)?,
+                        self.inferencer.infer_writes(txn)?,
+                    ))
+                })
+                .collect::<AResult<Vec<_>>>()
+            {
+                Ok(res) => res,
+                // Inferencer passed in by user failed to get the read/writeset of a transaction,
+                // abort parallel execution.
+                Err(_) => return Err(Error::InferencerError),
+            }
+        };
+
+        // Use write analysis result to construct placeholders.
+        let path_version_tuples: Vec<(T::Key, usize)> = infer_result
+            .par_iter()
+            .enumerate()
+            .with_min_len(chunks_size)
+            .fold(Vec::new, |mut acc, (idx, (_, txn_writes))| {
+                acc.extend(txn_writes.clone().into_iter().map(|ap| (ap, idx)));
+                acc
+            })
+            .flatten()
+            .collect();
+
+        let (versioned_data_cache, max_dependency_level) =
+            MVHashMap::new_from_parallel(path_version_tuples);
+        let outcomes = OutcomeArray::new(num_txns);
+
+        let scheduler = Arc::new(Scheduler::new(num_txns));
+
+        scope(|s| {
+            // How many threads to use?
+            let compute_cpus = min(1 + (num_txns / 50), self.num_cpus - 1); // Ensure we have at least 50 tx per thread.
+            let compute_cpus = min(num_txns / max_dependency_level, compute_cpus); // Ensure we do not higher rate of conflict than concurrency.
+
+            for _ in 0..(compute_cpus) {
+                s.spawn(|_| {
+                    let scheduler = Arc::clone(&scheduler);
+                    // Make a new executor per thread.
+                    let task = E::init(task_initial_arguments);
+
+                    while let Some(idx) = scheduler.next_txn_to_execute() {
+                        let txn = &signature_verified_block[idx];
+                        let (reads, writes) = &infer_result[idx];
+
+                        // If the txn has unresolved dependency, adds the txn to deps_mapping of its dependency (only the first one) and continue
+                        if reads
+                            .iter()
+                            .any(|k| match versioned_data_cache.read(k, idx) {
+                                Err(Some(dep_id)) => scheduler.add_dependency(idx, dep_id),
+                                Ok(_) | Err(None) => false,
+                            })
+                        {
+                            // This causes a PAUSE on an x64 arch, and takes 140 cycles. Allows other
+                            // core to take resources and better HT.
+                            ::std::hint::spin_loop();
```

Comment:
I don't think hint is useful here, we need it when a thread doesn't find work, but there is potentially more work left (just not dependency-cleared atm).